### PR TITLE
[#312] support for Gitlab

### PR DESCRIPTION
**Pull request by mbao01**
Short checklist:

  - ✅ Decouple GitHub from class implementations
  - ✅ Create "abstract" class (like a blueprint/interface) to integrate new VCS into project.
  - ✅ Create GitHub client and helper classes
  - ✅ Test GitHub client and helper classes
  - ✅ Live test GitHub implementation and should work as previously did
  - ✅  Update application unit tests for refactor and Github integration
  - ✅ Create Gitlab client and helper classes
  - ✅ Test Gitlab client and helper classes
  - ✅ Live test Gitlab implementation
  - ✅ Add application unit tests for JIRA integration
  - ✅ Handle errors appropriately
  - ✅ Live test application for Gitlab and Github

### CODE REVIEW

**Comment by mbao01** [comment]
@yegor256 please review. This will be a huge PR so it is better to review incrementally. Thanks

**Comment by yegor256** [comment]
@mbao01 still working on this?

**Comment by yegor256** [comment]
@mbao01 what's up with this?

**Comment by 0crat** [comment]
@yegor256[/z](https://www.0crat.com/u/yegor256) everybody who has role `REV` is banned at [#315](https://github.com/yegor256/0pdd/issues/315); I won't be able to assign anyone automatically; consider assigning someone manually (as in [§19](http://www.zerocracy.com/policy.html#19)), or invite more people (as in [§51](http://www.zerocracy.com/policy.html#51)), or remove the job from the scope (as in [§14](http://www.zerocracy.com/policy.html#14))

**Comment by mbao01** [comment]
@yegor256 please review

**Comment by 0crat** [comment]
@yegor256[/z](https://www.0crat.com/u/yegor256) everybody who has role `REV` is banned at [#315](https://github.com/yegor256/0pdd/issues/315); I won't be able to assign anyone automatically; consider assigning someone manually (as in [§19](http://www.zerocracy.com/policy.html#19)), or invite more people (as in [§51](http://www.zerocracy.com/policy.html#51)), or remove the job from the scope (as in [§14](http://www.zerocracy.com/policy.html#14))

File: Rakefile [hunk 1]
```diff
@@ -76,7 +76,7 @@ end
 desc 'Sleep endlessly after the start of DynamoDB Local server'
 task :sleep do
   loop do
-    sleep(5)
+    sleep(3600)
```
Comment (yegor256) [inline Rakefile:79]: @mbao01 this doesn't sound nice :)

File: dynamodb-local/tables/0pdd-events.json [hunk 2]
```diff
@@ -2,7 +2,7 @@
   "TableName": "0pdd-events",
   "AttributeDefinitions": [
     {
-      "AttributeName": "repo",
+      "AttributeName": "id",
```
Comment (yegor256) [inline dynamodb-local/tables/0pdd-events.json:5]: @mbao01 this is dangerous, we already have AWS Amazon scheman and won't change it. Why changing the name of the attribute?

File: objects/clients/gitlab.rb [hunk 3]
```diff
@@ -18,19 +18,34 @@
...
-
-  def all
-    @github.organization_memberships(state: 'pending').collect do |membership|
-      GithubOrganizationInvitation.new(membership, @github)
+class GitlabClient
+  def self.new(config = {})
+    begin
+      client = if config['testing']
+        require_relative '../../test/fake_gitlab'
+        FakeGitlab.new # TODO:: implement fake gitlab
```
Comment (yegor256) [inline objects/clients/gitlab.rb:32]: @mbao01 maybe it's better to use PDD `todo` marker?

File: objects/log.rb [hunk 4]
```diff
@@ -18,24 +18,35 @@
...
 # Log.
 #
 class Log
-  def initialize(aws, repo)
+  def initialize(dynamo, repo, vcs_name = 'github')
+    @dynamo = dynamo
     @repo = repo
-    @aws = aws
+    @vcs_name = (vcs_name || 'github').downcase
+    # TODO: Must perform seamless migration of existing DB
```
Comment (yegor256) [inline objects/log.rb:35]: @mbao01 why not PDD puzzle here?

File: objects/puzzles.rb [hunk 5]
```diff
@@ -64,35 +64,44 @@ def group(xml)
   end
 
   def expose(xml, tickets)
+    exceptions = [] # TODO: Send exceptions to the user's email or admin
```
Comment (yegor256) [inline objects/puzzles.rb:67]: @mbao01 why not PDD puzzle?

File: objects/puzzles.rb [hunk 6]
```diff
@@ -64,35 +64,44 @@ def group(xml)
...
+        puzzles = xml.xpath(
+          '//puzzle[@alive="false" and issue
+          and issue != "unknown" and not(issue/@closed)' +
+          seen.map { |i| "and id != '#{i}'" }.join(' ') + ']'
+        )
+        break if puzzles.empty?
+        puzzle = puzzles[0]
+        puzzle.search('issue')[0]['closed'] = Time.now.iso8601 if tickets.close(puzzle)
+        save(xml)
+      rescue Exception => e
```
Comment (yegor256) [inline objects/puzzles.rb:80]: @mbao01 I think we have a special decorator for catching exceptions here. If we don't, let's make it. Now you mix together two features: fetching puzzles and logging exceptions.

File: objects/vcs/github.rb [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+require 'octokit'
+require_relative 'base'
+require_relative '../git_repo'
+require_relative '../clients/gitlab'
+
+#
+# Github helper
+# API: https://github.com/NARKOZ/gitlab
+#
+class GithubHelper
```
Comment (yegor256) [inline objects/vcs/github.rb:30]: @mbao01 read this: https://www.yegor256.com/2015/03/09/objects-end-with-er.html

File: objects/vcs/gitlab.rb [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+require_relative 'base'
+require_relative '../git_repo'
+require_relative '../clients/gitlab'
+
+#
+# Gitlab client
+# API: https://github.com/NARKOZ/gitlab
+#
+class GitlabHelper
+  include VCS
```
Comment (yegor256) [inline objects/vcs/gitlab.rb:31]: @mbao01 read this: https://www.yegor256.com/2017/03/07/traits-and-mixins.html

File: request.http [hunk 9]
```diff
@@ -0,0 +1,272 @@
+POST http://localhost:4567/hook/github
```
Comment (yegor256) [inline request.http:1]: @mbao01 what this file is for? garbage?

**Comment by mbao01** [comment]
@yegor256 I resolved the comments. We will break this down into smaller PRs.

**Comment by mbao01** [comment]
Fixed in #356

**Comment by 0crat** [comment]
Pull request [#315](https://github.com/yegor256/0pdd/issues/315) was not merged, no payment for `ARC`, see [§28](http://www.zerocracy.com/policy.html#28)

**Pull request closed without merging**